Adding docs on using tokio task_local to share state

Gaelik-git opened the pull request:
I stumbled upon this [discussion](https://github.com/tokio-rs/axum/discussions/2631) for using tokio task_local to share state between handlers and middleware.

Is it okay to add some docs, would you prefer to have some test or example to show the behaviour ?

jplatte commented on axum/src/docs/middleware.md:
```diff
@@ -535,6 +535,60 @@ let app = Router::new()
 automatically moved to response extensions. You need to manually do that for the
 extensions you need.
 
+State can also be passed from middleware to handlers using [tokio task_local]:
```
```suggestion
State can also be passed from middleware to handlers using
[tokio's `task_local` macro](https://docs.rs/tokio/1/tokio/macro.task_local.html):
```

jplatte commented on axum/src/docs/middleware.md:
```diff
@@ -535,6 +535,60 @@ let app = Router::new()
+use tokio::task_local;
+
+#[derive(Clone)]
+struct CurrentUser {/* ... */}
```
```suggestion
struct CurrentUser {
    // ...
}
```

jplatte commented on axum/src/docs/middleware.md:
```diff
@@ -535,6 +535,60 @@ let app = Router::new()
+}
+
+async fn auth(req: Request, next: Next) -> Result<Response, StatusCode> {
+    let auth_header = req
+        .headers()
+        .get(http::header::AUTHORIZATION)
+        .and_then(|header| header.to_str().ok());
+
+    let auth_header = if let Some(auth_header) = auth_header {
+        auth_header
+    } else {
+        return Err(StatusCode::UNAUTHORIZED);
+    };
```
```suggestion
    let auth_header = req
        .headers()
        .get(http::header::AUTHORIZATION)
        .and_then(|header| header.to_str().ok())
        .ok_or(StatusCode::UNAUTHORIZED)?;
```

jplatte reviewed (commented):
Thanks for the PR! I think it makes sense to showcase this in the docs, but in addition to the small improvement suggestions below, there are two more general things:

- I think this documentation belongs in the main documentation, in the "Sharing state with handlers" section
- To me, the real important aspect about this is that unlike other state sharing mechanisms, this one actually also works to share state with `IntoResponse` impls; I think the docs should call this out

jplatte commented:
Hey @Gaelik-git, are you still interested in getting back to this?

Gaelik-git commented:
Yep sorry for the late response, I moved and modified the example for it to include the IntoReponse implementation
But my example does not make much sense I think, devs will still get the point I think but if you have other sugestion do not hesitate

jplatte commented on axum/src/lib.rs:
```diff
@@ -293,6 +293,68 @@
 //!
+//! ## Using [tokio's `task_local` macro](https://docs.rs/tokio/1/tokio/macro.task_local.html):
+//!
+//! This allows to share state with IntoResponse implementations.
```
```suggestion
//! This allows to share state with `IntoResponse` implementations.
```

jplatte commented on axum/src/lib.rs:
```diff
@@ -293,6 +293,68 @@
+//! let app: Router = Router::new()
+//!     .route("/", get(handler))
+//!     .route_layer(middleware::from_fn(auth));
+//!
```
The codeblock will look better without an empty line at the end.

```suggestion
```

jplatte reviewed (approved):
Thanks!

Gaelik-git commented:
glad I could contribute
